src/mirage_protocols.mli

```diff
@@ -383,13 +399,35 @@ module type TCP = sig
       Note that this API will change in a future revision to be a
       per-flow attribute instead of a separately exposed function. *)
 
-  val create_connection: t -> ipaddr * int -> (flow, error) result io
-  (** [create_connection t (addr,port)] opens a TCPv4 connection
-      to the specified endpoint. *)
+  val create_connection: ?keepalive:Keepalive.t -> t -> ipaddr * int -> (flow, error) result io
+  (** [create_connection ~keepalive t (addr,port)] opens a TCPv4 connection
+      to the specified endpoint.
+
+      If the optional argument [?keepalive] is provided then TCP keep-alive
+      messages will be sent to the server when the connection is idle. If
+      no responses are received then eventually the connection will be disconnected:
+      [read] will return [Ok `Eof] and write will return [Error `Closed] *)
+
+  type listener = {
+    process: flow -> unit io; (** process a connected flow *)
+    keepalive: Keepalive.t option; (** optional TCP keepalive configuration *)
+  }
+  (** A TCP listener on a particular port *)
+
+  val input: t -> listeners:(int -> listener option) -> ipinput
+  (** [input t listeners] returns an input function continuation to be
+      passed to the underlying {!IP} stack.
+
+      When the stack receives a TCP SYN (i.e. a connection request) to a
+      particular [port], it will evaluate [listeners port]:
+
+      - If [listeners port] is [None], the input function will return an RST
+        to refuse the connection.
+      - If [listeners port] is [Some listener] then the connection will be
+        accepted and the resulting flow will be processed by [listener.process].
+        If [listener.keepalive] is [Some configuration] then the TCP keep-alive
+        [configuration] will be applied before calling [listener.process].
+  *)
+
 
-  val input: t -> listeners:(int -> callback option) -> ipinput
```

Comment:
as you don't use the `callback` type here maybe we should just delete that type?